### ISSUE TITLE
Request for a different type of example of abort steps

### ISSUE THREAD

**Issue by domenic** [whatwg/dom#1031]
Might you folks be able to point at a more detailed example of how one's "[abort steps](https://dom.spec.whatwg.org/#abortcontroller-api-integration)" might signal to one's overall algorithm that it needs to "Stop doing amazing things." ? 

I am guessing that one's abort steps could set a "stop doing things" flag, and one's overall algorithm's in-parallel steps could check that flag now-and-then and if true then one's spec would simply say to "abort these alg steps" (it wouldn't need to do the promise rejection at that point because the "abort steps" (per [abortcontroller-api-integration](https://dom.spec.whatwg.org/#abortcontroller-api-integration)) take care of it) ....?

_Originally posted by @equalsJeffH in https://github.com/whatwg/dom/issues/1030#issuecomment-965902194_

**Comment by domenic** [comment]
> I am guessing that one's abort steps could set a "stop doing things" flag, and one's overall algorithm's in-parallel steps could check that flag now-and-then

Yes, that sounds reasonable.

> it wouldn't need to do the promise rejection at that point because the "abort steps" (per abortcontroller-api-integration) take care of it)

Probably you want to delay the promise rejection until you've actually confirmed the abort went through. I.e. in such a flag-based approach you would not want to follow the template linked there, but instead queue a task from in parallel to reject the promise.

Otherwise you could, for example, have a race, where on the main thread you set the flag, but in parallel you've finished the operation, and so setting the flag does nothing. In that case you would not want to reject the promise.

**Comment by equalsJeffH** [comment]
> instead queue a task from in parallel to reject the promise

but of course, doh!

How does the below look?
Is there an approach other than a flag-based one?

```
1. Let p be a new promise.

2. If options["signal"] member is present, then:

    1. Let signal be options["signal"].
    
    2. Let abortedFlag be false.
       Let algCompleteFlag be false.

    3. If signal is aborted, then reject p 
       with signal’s abort reason and return p.

    4. Add the following abort steps to signal:

        1. Set abortedFlag to true.

3. Run these steps in parallel:

    1. Do some amazing things, which includes checking now-and-then
       for abortedFlag being true. 
       
        1. If abortedFlag is found to be true, then:
        
            1. Queue a task on <an appropriate task source> to run the following substeps:
        
                1. If algCompleteFlag is false, then:
                
                    1. Reject p with signal’s abort reason.
                
                    2. Abort these in-parallel steps.

    2. Let amazingResult be the result of doing the amazing things.

    3. Resolve p with amazingResult.

    4. Set algCompleteFlag to true.

4. Return p.

```

**Comment by domenic** [comment]
Looks pretty good to me! Only one issue, not related to the aborting stuff: don't forget to queue a task for "resolve" too; and, try these days to use "queue a global task" instead of "queue a task" for precision.

I think the alternative to the flag approach would be when the in-parallel action is vague enough that specs sum it up in a single step, and so you can say something similarly-vague like "interrupt the process". E.g. something like

1. Let global be this's relevant global object.

1. Let p be a new promise.

1. If options["signal"] member is present, then:

    1. Let signal be options["signal"].
    
    3. If signal is aborted, then reject p 
       with signal’s abort reason and return p.

    4. Add the following abort steps to signal:

        1. Abort the process of reading from the I/O device.

1. In parallel, read from the IO device. When this completes (in any fashion), queue a global task on the X task source given global to perform the following steps:

    1. If reading completed successfully, resolve p with the result of creating an ArrayBuffer from the bytes read.

    1. If reading failed due to I/O issues, reject p with a "OperationError" DOMException.

    1. If reading was aborted, reject p with signal's abort reason.

4. Return p.

If your in-parallel action has multiple interesting steps, with well-defined points at which you would check the aborted flag, then being precise about that like you outlined is better.

**Comment by equalsJeffH** [comment]
> ... don't forget to queue a task for "resolve" too; ...

Sorry, I don't understand why/how to do that, and I note you did not do such in [your example](https://github.com/whatwg/dom/issues/1031#issuecomment-966516722)?

**Comment by domenic** [comment]
> I  don't understand why/how to do that, 

Your step 3 should be "Queue a global task on the X task source given Y to Resolve p with amazingResult." Right now it does not queue a task, it calls "resolve" directly from in parallel steps.

> I note you did not do such in your example?

I did. My step 4.i resolves the promise, and is nested under "queue a global task on the X task source given global to perform the following steps".

**Comment by annevk** [comment]
We probably want to clarify the specification here, but do we really want folks to postpone rejecting the promise until the abort has succeeded? That doesn't seem to match the early synchronous abort check we also recommend.

I think we want to signal abort straightaway and in parallel do our best to also abort the ongoing operation.

We also need to update the example here to queue a task.

**Comment by domfarolino** [comment]
I guess this is a question of whether we want Promise-rejection-as-a-result-of-abortion to be a good signal for whether the underlying operation truly got canceled, or to a lesser extent had side-effects, right?

E.g., if I do `const p = fetch("/foo", {method: "POST", signal, ...})`, and then abort the fetch sometime later, should:
 1. `p` only reject if were able to meaningfully cancel the underlying operation (fetch), and in this case I can be assured that the POST didn't touch anything, or..
 2. `p` reject immediately, and cancel the fetch best-effort

Fetch does (2), but at least one other user of AbortController/Signal I'm aware of does (1), and I'm commenting here as a result of reviewing code in Chromium that is leaning towards doing (1) too.

A related question is: should it be possible for promise-returning APIs to ever resolve after we `abort()` the underlying operation? If we prefer (1) above, then certainly. If (2), then I'm not sure it's possible to ever see the promise resolve after calling abort(). I'd expect abort() to immediately reject the promise (something we could add in the normative requirements in https://dom.spec.whatwg.org/#abortcontroller-api-integration if we wanted), so that in the next microtask the promise is indeed rejected. The only way to see a promise resolve after `abort()` is if in the same task that `abort()` is called, `Reject()` is called just before it.

Between (1) or (2), I am not sure which is a better model or which way most usages of AbortController/Signal lean. I am not even sure if we want to prescribe a one-size-fits-all guideline in https://dom.spec.whatwg.org/#abortcontroller-api-integration, or just leave it up to each consumer API.

**Comment by annevk** [comment]
So thinking about this more it would probably be more useful for API consumers if we only signaled abortion if abortion succeeded. So we set the aborted flag, and if the in parallel operation indeed had not completed when it observes the set aborted flag it aborts, which then results in a task that rejects the promise.

I think we could probably also still change `fetch()` to do that, although it would be an observable change.

I do think we want this to be consistent across "in parallel" APIs. Having to learn it for each API and account for it would be rather frustrating.

(Removing "clarification" is this is becoming a bit more than that.)

**Comment by annevk** [comment]
I ended up creating a half dozen PRs against Fetch to try to resolve this, but there are quite a few tricky aspects with actually aborting the underlying operation first (or at all). So in the end I settled for some internal notes so we can hopefully solve this in the future: https://github.com/whatwg/fetch/pull/1514. (The other PRs did meaningfully improve Fetch so definitely worth it overall I think.)

Now In general I think we should aim for aborting the underlying operation first and once that is done exposing it to script. That's a logical order and where we can do it we should aim for it.
